midi/midi/encode.py

```diff
@@ -4,227 +4,245 @@
+    """
+    translates a multi-dimensional array into a MIDI file
+    """
+    tempos, accuracy, midi_file = prepare_meta_file(tempos, grid_accuracy, event_lengths)
+    if data.ndim == 2:
+        midi_file.tracks.append(get_messages_from_standard_2d_input(data, 0, tempos, accuracy, join_notes,
+                                                                    use_sequences, use_velocities, event_lengths))
 
+    elif data.ndim == 3:
+        for i in range(data.shape[0]):  # channels are limited to 16 in MIDI 1.0
+            midi_file.tracks.append(get_messages_from_standard_2d_input(data[i], i % 16, tempos, accuracy, join_notes,
+                                                                        use_sequences, use_velocities, event_lengths))
 
-def generate_file_from_tonal_features(input_array, output_filename, accuracy):   # TODO: check if even and > 0
+    else:
+        raise TypeError('input array must have 2 or 3 dimensions')
+
+    midi_file.save(output_path)
+
+
+def get_file_from_music21_features(data, output_path, use_tonal_features, join_notes, grid_accuracy=GRID_ACCURACY):
     """
     translates a music_21.py output matrix of tonal features into a MIDI file
-
-    :param input_array:
-    :param output_filename:
-    :param accuracy:
-    :return:
     """
-    if input_array.ndim != 2:
+    if data.ndim != 2:
         raise TypeError('input array must have exactly 2 dimensions')
 
+    if data.shape[1] <= 2:
+        raise ValueError('number of MIDI tracks with notes\' messages must be at least 1')
+
+    if use_tonal_features and data.shape[1] % 2 == 1:
+        raise ValueError('odd number of tracks - each tonal track must be provided with two features')
+
+    event_lengths = []
     tempos = []
     array = []
 
-    for feature_index in range(int(input_array.shape[1] / 2 - 1)):
+    for feature_index in range(int(data.shape[1] / 2 - 1)):
         track = []
-        for event_index in range(input_array.shape[0]):
+        for event_index in range(data.shape[0]):
             notes = [False] * 128
-            if input_array[event_index][2 * feature_index + 1] != 0:
-                feature = round((input_array[event_index][2 * feature_index] * 11 - 1) * 12 +
-                                input_array[event_index][2 * feature_index + 1] * 12 - 1)
-                notes[feature] = True
 
-            length = round(1 / input_array[event_index][-2])
-            track.extend([notes] * length)
+            if use_tonal_features:
+                if data[event_index][2 * feature_index + 1] != 0:  # ignore zero-length events
+                    feature = round((data[event_index][2 * feature_index] * 11 - 1) * 12 +
+                                    data[event_index][2 * feature_index + 1] * 12 - 1)
+                    notes[feature] = True
+
+            else:
+                feature = round(data[event_index][feature_index] * 128 - 1)
```

Comment:
Would be nice if those magic numbers would be enclosed in some constants.

Reply:
those magic numbers are number of notes in MIDI, don't be an object-oriented pussy

Reply:
Constants have nothing to do with OOP. As a reader of this code I have no idea what these numbers mean, making it hard for me to edit this code.

Reply:
You should not parameterize values which may not be changed. However, I can add comments to these calculations

Reply:
descriptive comments added in ed79d5dcd403e9b7edea21c6781f58840aead36f

Reply:
Its not parametrization, these are constants